src/coreclr/jit/codegencommon.cpp

```diff
@@ -4024,9 +4024,15 @@ void CodeGen::genZeroInitFrame(int untrLclHi, int untrLclLo, regNumber initReg,
                 continue;
             }
 
-            // TODO-Review: I'm not sure that we're correctly handling the mustInit case for
-            // partially-enregistered vars in the case where we don't use a block init.
-            noway_assert(varDsc->lvIsInReg() || varDsc->lvOnFrame);
+            // Locals that are (only) in registers to begin with do not need
+            // their stack home zeroed. Their register will be zeroed later in
+            // the prolog.
+            if (varDsc->lvIsInReg() && !varDsc->lvLiveInOutOfHndlr)
+            {
+                continue;
+            }
+
+            noway_assert(varDsc->lvOnFrame);
```

Comment:
In `genCheckUseBlockInit` it looks like the main loop only increases `initStkLclCnt` when `varDsc->lvIsOnFrame`.

So couldn't we use that as an early out detection mechanism? Basically:
```cpp
if (!varDsc->lvMustInit || !varDsc->lvOnFrame)
{
    continue;
}
```

There's then a secondary handler that increments for spill temps containing pointers, but that looks to be independent of this main loop in `genZeroInitFrame` and is handled in its own subsequent loop.

Reply:
`genCheckUseBlockInit` will set `lvMustInit`, so I don't think we need more checks here... the logic is that if `genCheckUseBlockInit` left `lvMustInit` as 1 and it isn't a register argument, then it must be on frame as otherwise we wouldn't need to init it.